pkg/queue/concurrency_state.go

```diff
@@ -93,33 +104,84 @@ func ConcurrencyStateHandler(logger *zap.SugaredLogger, h http.Handler, pause, r
 	}
 }
 
+// handleStateRequestError handles different error code
+func handleStateRequestError(errCode int8, requestHandler, deleteFunc func() (int8, error)) {
+	if errCode == responseStatusConflictError {
+		// nothing should be done, just ignore
+	}
+	if errCode == internalError {
+		os.Exit(1)
+	}
+	if errCode == responseExecError {
+		failedTimes := 0
+		for failedTimes < FreezeMaxRetryTimes {
+			errCode, _ := requestHandler()
+			if errCode == responseExecError {
+				failedTimes++
+				time.Sleep(time.Millisecond * 200)
+			}
+			if errCode == internalError {
+				os.Exit(1)
+			}
+			if errCode == responseStatusConflictError || errCode == noError {
+				break
+			}
+		}
+		if failedTimes >= FreezeMaxRetryTimes {
+			// Relaunch this pod, the way is: the runtime will delete all containers of this pod
+			errCode, error := deleteFunc()
+			// if the QP is deleted, this will not be executed
+			for error != nil {
+				// the error code type here could be: internalError, responseExecError(failed to exec delete command)
+				if errCode == internalError {
+					os.Exit(1)
+				} else {
+					_, error = deleteFunc()
+				}
+			}
+		}
+	}
+}
+
 // concurrencyStateRequest sends a request to the concurrency state endpoint.
-func concurrencyStateRequest(endpoint string, action string) func() error {
-	return func() error {
+func concurrencyStateRequest(endpoint string, action string) func() (int8, error) {
+	return func() (int8, error) {
 		bodyText := fmt.Sprintf(`{ "action": %q }`, action)
 		body := bytes.NewBufferString(bodyText)
 		req, err := http.NewRequest(http.MethodPost, endpoint, body)
 		if err != nil {
-			return fmt.Errorf("unable to create request: %w", err)
+			return internalError, fmt.Errorf("unable to create request: %w", err)
 		}
 		req.Header.Add("Token", "nil") // TODO: use serviceaccountToken from projected volume
 		resp, err := http.DefaultClient.Do(req)
 		if err != nil {
-			return fmt.Errorf("unable to post request: %w", err)
+			return internalError, fmt.Errorf("unable to post request: %w", err)
+		}
+		if resp.StatusCode == http.StatusConflict {
```

Comment:
I think I'd worry about more granular response codes and conflicts in a follow-on PR and focus this PR on just retrying (all) failed requests, regardless the status code. That way it'll be a bit easier to review the first version of this, and we'll end up with two smaller commits.

Reply:
I will change it.